### PR TITLE
[PLUGIN-650] Fixed: DLP Plugins do not support templates from other projects

### PR DESCRIPTION
**Pull request by flakrimjusufi**
DLP Plugins do not support templates from other projects
JIRA Ticket: https://cdap.atlassian.net/browse/PLUGIN-650

### CODE REVIEW

File: src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java [hunk 1]
```diff
@@ -79,12 +86,32 @@
...
+      String.format("projects/%s/inspectTemplates/%s", getProject(), templateId);
+  }
 
   public void validate(FailureCollector collector, Schema inputSchema) {
     if (customTemplateEnabled) {
-      if (!containsMacro("templateId") && templateId == null) {
-        collector.addFailure("Must specify template ID in order to use custom template", "")
-          .withConfigProperty("templateId");
+      if (!containsMacro(CUSTOM_TEMPLATE_PATH_NAME) && Strings.isNullOrEmpty(CUSTOM_TEMPLATE_PATH_NAME)) {
+        if (!containsMacro(TEMPLATE_ID_NAME) && Strings.isNullOrEmpty(TEMPLATE_ID_NAME)) {
```
Comment (ardian4) [inline src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java]: Can we remove nested `if` condition and just append the first `if` ?

File: src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java [hunk 2]
```diff
@@ -79,12 +86,32 @@
...
+          collector.addFailure("Must specify template ID in order to use custom template", "")
+            .withConfigProperty(TEMPLATE_ID_NAME);
+        }
+      }
+      if (!containsMacro(TEMPLATE_ID_NAME) && Strings.isNullOrEmpty(TEMPLATE_ID_NAME)) {
```
Comment (ardian4) [inline src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java]: Is this condition same as above condition?

File: src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java [hunk 3]
```diff
@@ -79,12 +86,32 @@
...
+        if (!containsMacro(CUSTOM_TEMPLATE_PATH_NAME) && Strings.isNullOrEmpty(CUSTOM_TEMPLATE_PATH_NAME)) {
+          collector.addFailure("Must specify custom Template Path in order to use custom template", "")
+            .withConfigProperty(CUSTOM_TEMPLATE_PATH_NAME);
+        }
+      }
+      if (!Strings.isNullOrEmpty(templateId) && !Strings.isNullOrEmpty(customTemplatePath)) {
```
Comment (ardian4) [inline src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java]: Can you check here for macro also?

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java [hunk 4]
```diff
@@ -334,8 +350,35 @@ public int onErrorHandling() {
...
+     */
+    public String getCustomTemplate() {
+      return Strings.isNullOrEmpty(templateId) ? customTemplatePath :
+        String.format("projects/%s/inspectTemplates/%s", getProject(), templateId);
+    }
 
     public void validate(FailureCollector collector, Schema inputSchema) {
+      if (customTemplateEnabled) {
+        if (!containsMacro(CUSTOM_TEMPLATE_PATH_NAME) && Strings.isNullOrEmpty(customTemplatePath)) {
+          if (!containsMacro(TEMPLATE_ID_NAME) && Strings.isNullOrEmpty(templateId)) {
```
Comment (ardian4) [inline src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: Can you move the nested condition form nested `if` to first `if`?

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java [hunk 5]
```diff
@@ -334,8 +350,35 @@ public int onErrorHandling() {
...
+            collector.addFailure("Must specify one of template Id or Custom Template Path", "")
+              .withConfigProperty(CUSTOM_TEMPLATE_ENABLED_NAME);
+          }
+        }
+        if (!containsMacro(TEMPLATE_ID_NAME) && Strings.isNullOrEmpty(templateId)) {
```
Comment (ardian4) [inline src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: Is this part of the code same as above check?

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java [hunk 6]
```diff
@@ -334,8 +350,35 @@ public int onErrorHandling() {
...
+          if (!containsMacro(CUSTOM_TEMPLATE_PATH_NAME) && Strings.isNullOrEmpty(customTemplatePath)) {
+            collector.addFailure("Must specify one of template Id or Custom Template Path", "")
+              .withConfigProperty(CUSTOM_TEMPLATE_ENABLED_NAME);
+          }
+        }
+        if (!Strings.isNullOrEmpty(templateId) && !Strings.isNullOrEmpty(customTemplatePath)) {
```
Comment (ardian4) [inline src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: Can you check here those properties are macro?

File: src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java (hunk 1 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java]: PR updated.

File: src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java (hunk 3 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java]: PR updated.

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java (hunk 4 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java:366]: PR updated.

File: src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java (hunk 2 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java]: No. This condition is the opposite of above.

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java (hunk 4 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: PR updated.

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java (hunk 6 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: PR updated.

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java (hunk 5 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: This check is the opposite of the one above.

Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: This condition is the opposite of above. It checks whether both of custom template fields are filled. I simplified the conditions in the last updated PR.

File: src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java (hunk 2 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/DLPTransformPluginConfig.java]: This condition is the opposite of above. It checks whether both of custom template fields are filled (not empty). I simplified the conditions in the last updated PR.

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java [hunk 7]
```diff
@@ -334,8 +350,29 @@ public int onErrorHandling() {
       return 1;
     }
 
+    /**
+     * Gets the custom template specified by user which can be either a templateId or a full custom template path
+     *
+     * @return String of templateId or custom template path
+     */
+    public String getCustomTemplate() {
+      return Strings.isNullOrEmpty(templateId) ? customTemplatePath :
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java:359]: Please add documentation for the new customTemplatePath config.

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java [hunk 8]
```diff
@@ -334,8 +350,29 @@ public int onErrorHandling() {
...
+        String.format("projects/%s/inspectTemplates/%s", getProject(), templateId);
+    }
 
     public void validate(FailureCollector collector, Schema inputSchema) {
+      if (customTemplateEnabled) {
+        if (!containsMacro(TEMPLATE_ID_NAME) && !containsMacro(CUSTOM_TEMPLATE_PATH_NAME) &&
+          Strings.isNullOrEmpty(templateId) && Strings.isNullOrEmpty(customTemplatePath)) {
+          collector.addFailure("Must specify one of the custom template fields", "")
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: collector.addFailure("Custom template fields are not specified", "Must specify one of template id or template path")

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java [hunk 9]
```diff
@@ -334,8 +350,29 @@ public int onErrorHandling() {
...
+            .withConfigProperty(TEMPLATE_ID_NAME).withConfigProperty(CUSTOM_TEMPLATE_PATH_NAME);
+        }
+        if (!containsMacro(TEMPLATE_ID_NAME) && !containsMacro(CUSTOM_TEMPLATE_PATH_NAME) &&
+          !Strings.isNullOrEmpty(templateId) && !Strings.isNullOrEmpty(customTemplatePath)) {
+          collector.addFailure("Must specify only one of the custom template fields", "")
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: collector.addFailure("Both template id and template path are specified", "Must specify only one of template id or template path")

File: widgets/SensitiveRecordFilter-splittertransform.json [hunk 10]
```diff
@@ -199,6 +223,19 @@
           "name": "dlp-tls-enabled"
         }
       ]
+    },
+    {
+      "name": "Filter for simple condition objects - 1",
```
Comment (rmstar) [inline widgets/SensitiveRecordFilter-splittertransform.json:228]: what is `Filter for simple condition objects - 1`?

File: src/test/java/io/cdap/plugin/dlp/UtilsTest.java [hunk 11]
```diff
@@ -65,4 +67,42 @@ public void testTableFromStructuredRecord() throws Exception {
...
   }
+
+  @Test(expected = ValidationException.class)
+  public void testEmptyCustomTemplate() {
+    DLPTransformPluginConfig config = new DLPTransformPluginConfig();
+    config.customTemplateEnabled = true;
+    config.templateId = null;
+    MockFailureCollector collector = new MockFailureCollector("customTemplateEnabledAndEmpty");
+    config.validate(collector, null);
+    collector.getOrThrowException();
```
Comment (rmstar) [inline src/test/java/io/cdap/plugin/dlp/UtilsTest.java]: Check the validation failure details similar to what you've done in `testBothCustomTemplateEnabled()`

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java (hunk 7 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java:359]: PR updated.

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java (hunk 8 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: PR updated accordingly.

File: src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java (hunk 9 above)
Comment (flakrimjusufi) [reply src/main/java/io/cdap/plugin/dlp/SensitiveRecordFilter.java]: PR updated accordingly.

File: widgets/SensitiveRecordFilter-splittertransform.json (hunk 10 above)
Comment (flakrimjusufi) [reply widgets/SensitiveRecordFilter-splittertransform.json:228]: We used the same name as it was specified in other widgets when `customTemplateEnabled` was present. 

https://github.com/data-integrations/dlp-plugins/blob/561e837d836800641bd6e21964230b7cf1bce0a1/widgets/SensitiveRecordDecrypt-transform.json#L468-L475

https://github.com/data-integrations/dlp-plugins/blob/561e837d836800641bd6e21964230b7cf1bce0a1/widgets/SensitiveRecordRedaction-transform.json#L724-L731

When `customTemplateEnabled` is NO (disabled),  you can't specify any template. I think that's the reason why it as named like this.

File: src/test/java/io/cdap/plugin/dlp/UtilsTest.java (hunk 11 above)
Comment (flakrimjusufi) [reply src/test/java/io/cdap/plugin/dlp/UtilsTest.java]: PR updated accordingly.

File: widgets/SensitiveRecordFilter-splittertransform.json (hunk 10 above)
Comment (rmstar) [reply widgets/SensitiveRecordFilter-splittertransform.json:228]: I see, thanks

File: docs/SensitiveRecordDecrypt-transform.md [hunk 12]
```diff
@@ -27,4 +27,8 @@ Metrics
 This plugin records three metrics:
 * `dlp.requests.count`: Total number of requests sent to Data Loss Prevention API
 * `dlp.requests.success`: Number of requests that were successfully processed by Data Loss Prevention API
-* `dlp.requests.fail`: Number of requests that failed
\ No newline at end of file
+* `dlp.requests.fail`: Number of requests that failed
+
+Custom Template Path
+-----------
+The availability to use a custom template path which is located in a different project other than the one specified in Project Id.
```
Comment (rmstar) [inline docs/SensitiveRecordDecrypt-transform.md]: nit: availability -> option. 
Same thing in SensitiveRecordFilter-splittertransform.md

**Review by rmstar — APPROVED** [review]
👍  please rebase on top of latest develop branch and squash commits

**Comment by flakrimjusufi** [comment]
>  please rebase on top of latest develop branch and squash commits

Commits squashed.

**Pull request merged**